# LINEAR REGRESSION PROJECT

## GOAL
the goal of this project is to create a Linear Regression Model, train it on the data with the optimal parameters using a grid search, and then evaluate the model's capabilities on a test set.

In [1]:
# import nessesary libraries and read in the cleaned dataset
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv("../DATA/AMES_Cleaned_DF.csv")

In [2]:
df

Unnamed: 0  MS SubClass_150  MS SubClass_160  ...  Mo Sold  Yr Sold  SalePrice
0              0                0                0  ...        5     2010     215000
1              1                0                0  ...        6     2010     105000
2              2                0                0  ...        6     2010     172000
3              3                0                0  ...        4     2010     244000
4              4                0                0  ...        3     2010     189900
...          ...              ...              ...  ...      ...      ...        ...
2920        2925                0                0  ...        3     2006     142500
2921        2926                0                0  ...        6     2006     131000
2922        2927                0                0  ...        7     2006     132000
2923        2928                0                0  ...        4     2006     170000
2924        2929                0                0  ...       11     2006     188000

[2925 rows x 275 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925 entries, 0 to 2924
Columns: 275 entries, Unnamed: 0 to SalePrice
dtypes: float64(11), int64(264)
memory usage: 6.1 MB


seperate the data into label (what we are trying to predict in this case Sale price) and Fetures (what we will be using to predict the label)

In [8]:
X = df.drop("SalePrice", axis=1)
y = df["SalePrice"]

In [10]:
X.head()

Unnamed: 0  MS SubClass_150  MS SubClass_160  ...  Misc Val  Mo Sold  Yr Sold
0           0                0                0  ...         0        5     2010
1           1                0                0  ...         0        6     2010
2           2                0                0  ...     12500        6     2010
3           3                0                0  ...         0        4     2010
4           4                0                0  ...         0        3     2010

[5 rows x 274 columns]

In [11]:
y.head()

0    215000
1    105000
2    172000
3    244000
4    189900
Name: SalePrice, dtype: int64

split up X and y into a training set and test set. 
we can use a smaller test size Since we will later be using a Grid Search strategy to get the best performing params for our model.

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=404)

In [13]:
X_train

Unnamed: 0  MS SubClass_150  MS SubClass_160  ...  Misc Val  Mo Sold  Yr Sold
780          780                0                0  ...         0        6     2009
2465        2470                0                0  ...         0        1     2006
2291        2296                0                0  ...         0        9     2007
483          483                0                0  ...         0        4     2009
2201        2205                0                0  ...         0        8     2007
...          ...              ...              ...  ...       ...      ...      ...
1935        1937                0                0  ...         0        8     2007
1012        1012                0                0  ...         0        9     2008
1206        1206                0                0  ...         0        4     2008
1898        1900                0                0  ...         0       10     2007
1788        1790                0                0  ...         0        8     2007

[2632 rows x 274 columns]

In [14]:
X_test

Unnamed: 0  MS SubClass_150  MS SubClass_160  ...  Misc Val  Mo Sold  Yr Sold
899          899                0                0  ...         0        8     2009
1875        1877                0                0  ...         0        6     2007
933          933                0                1  ...         0       10     2009
2444        2449                0                0  ...         0        6     2006
2609        2614                0                0  ...         0        7     2006
...          ...              ...              ...  ...       ...      ...      ...
1311        1311                0                0  ...         0        5     2008
2723        2728                0                0  ...         0        6     2006
2559        2564                0                0  ...         0        3     2006
1850        1852                0                0  ...         0        6     2007
1577        1579                0                0  ...         0        7     2008

[293 rows x 274 columns]

2632 rows from the training set + 293 from the test set = 2925 which is the total number of rows before spliting. 

The dataset features has a variety of scales and units. For optimal regression performance, scale the X features.

because we want to be able to test the models perdormance on truly unseen data, when scaling features becareful not to fit the test set to avoid any data leakage. 

In [15]:
# create an instance of the scaler fit only the training data and transform the features
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaled_X_train = scaler.fit_transform(X_train)
scaled_X_test = scaler.transform(X_test)

In [18]:
# create an instance of the model in this case elastic net.
from sklearn.linear_model import ElasticNet
base_elastic_model = ElasticNet(max_iter=1000000)

The Elastic Net model was chosen because it finds the best aproch between LASSO and Ridge Regression to be used based on the dataset. ihas two main parameters, alpha and the L1 ratio. Create a dictionary parameter grid of values for the ElasticNet.

In [17]:
param_grid = {'alpha':[0.1,1,5,10,50,100],
              'l1_ratio':[.1, .5, .7, .9, .95, .99, 1]}

create a GridSearchCV object and run a grid search for the best parameters for your model based on your scaled training data.

In [19]:
# create a gridsearch, fit the model on the training data
from sklearn.model_selection import GridSearchCV
grid_model = GridSearchCV(estimator=base_elastic_model,
                          param_grid=param_grid,
                          scoring='neg_mean_squared_error',
                          cv=5,
                          verbose=1)
grid_model.fit(scaled_X_train,y_train)

Fitting 5 folds for each of 42 candidates, totalling 210 fits


GridSearchCV(cv=5, estimator=ElasticNet(max_iter=1000000),
             param_grid={'alpha': [0.1, 1, 5, 10, 50, 100],
                         'l1_ratio': [0.1, 0.5, 0.7, 0.9, 0.95, 0.99, 1]},
             scoring='neg_mean_squared_error', verbose=1)

In [20]:
# you can find out which parameter the grid search found to be the best for the model
grid_model.best_params_

{'alpha': 100, 'l1_ratio': 1}

In [21]:
# get the models predictions
y_pred = grid_model.predict(scaled_X_test)

In [22]:
# use an error metric to Evaluate the model's performance on the unseen 10% scaled test set.
#for this i am using MAE and RMSE
from sklearn.metrics import mean_absolute_error,mean_squared_error
MAE = mean_absolute_error(y_test,y_pred)
MSE= mean_squared_error(y_test,y_pred)
RMSE = np.sqrt(MSE)

In [23]:
MAE

13413.118820082858

In [24]:
MSE

343787041.8668991

In [25]:
RMSE

18541.495135692243

to know if the prediction was good for the set of variables, we can check out the mean price of a house in the dataset

In [26]:
np.mean(df['SalePrice'])

180815.53743589742

we can see that we are not too off from the mean so it is a positve sign.
we could aso use LASSO Regression (Since Gridsearch leaned towards L1) to further simplyfy the daterset and try out other parameters to finetune the model then compare the predictions and errors between the models. 

we can then save and deploy this model to the real world to be used for future predictions

In [28]:
from joblib import dump
dump(base_elastic_model, "base_elastic_model.joblib")
dump(grid_model, 'sales_model.joblib') 

['sales_model.joblib']